File: scripts/gerar_icones.py
"""Gera icones PWA (bandeira IT) — apenas biblioteca padrao."""
import struct
import zlib
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent


def crc32(data: bytes) -> int:
    return zlib.crc32(data) & 0xFFFFFFFF


def chunk(tag: bytes, data: bytes) -> bytes:
    return struct.pack(">I", len(data)) + tag + data + struct.pack(">I", crc32(tag + data))


def png_rgba(size: int, draw) -> bytes:
    sig = b"\x89PNG\r\n\x1a\n"
    ihdr = struct.pack(">IIBBBBB", size, size, 8, 6, 0, 0, 0)
    rows = bytearray()
    for y in range(size):
        rows.append(0)
        for x in range(size):
            r, g, b, a = draw(x, y, size)
            rows.extend((r, g, b, a))
    comp = zlib.compress(bytes(rows), 9)
    return sig + chunk(b"IHDR", ihdr) + chunk(b"IDAT", comp) + chunk(b"IEND", b"")


def draw_icon(x, y, size):
    pad = int(size * 0.08)
    inner = size - pad * 2
    ix, iy = x - pad, y - pad
    if ix < 0 or iy < 0 or ix >= inner or iy >= inner:
        return (0, 98, 188, 255)
    third = inner / 3
    if ix < third:
        return (0, 146, 70, 255)
    if ix < third * 2:
        return (255, 255, 255, 255)
    return (206, 43, 55, 255)


def draw_maskable(x, y, size):
    cx = cy = size / 2
    r = size * 0.42
    if (x - cx + 0.5) ** 2 + (y - cy + 0.5) ** 2 > r * r:
        return (15, 20, 32, 255)
    return draw_icon(x, y, size)


for name, sz, fn in [
    ("icon-192.png", 192, draw_icon),
    ("icon-512.png", 512, draw_icon),
    ("icon-512-maskable.png", 512, draw_maskable),
]:
    out = ROOT / name
    out.write_bytes(png_rgba(sz, fn))
    print(f"OK {name} {sz}x{sz}")

File: scripts/gerar-icones.mjs
/**
 * Ícones PWA — ampulheta comum + faixa tricolor (comunidade, não oficial).
 * Requer: Node 18+ (sem dependências externas).
 */
import { writeFileSync } from 'fs';
import { deflateSync } from 'zlib';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';

const __dir = dirname(fileURLToPath(import.meta.url));
const root = join(__dir, '..');

const BG = [26, 22, 18];
const TERRA = [200, 90, 40];
const SAND = [243, 236, 227];
const VERDE = [0, 146, 70];
const BIANCO = [255, 255, 255];
const ROSSO = [206, 43, 55];

function crc32(buf) {
  let c = ~0;
  for (let i = 0; i < buf.length; i++) {
    c ^= buf[i];
    for (let k = 0; k < 8; k++) c = (c >>> 1) ^ (0xedb88320 & -(c & 1));
  }
  return ~c >>> 0;
}

function chunk(type, data) {
  const len = Buffer.alloc(4);
  len.writeUInt32BE(data.length);
  const t = Buffer.from(type);
  const crc = Buffer.alloc(4);
  crc.writeUInt32BE(crc32(Buffer.concat([t, data])));
  return Buffer.concat([len, t, data, crc]);
}

function pngSolid(size, draw) {
  const signature = Buffer.from([137, 80, 78, 71, 13, 10, 26, 10]);
  const ihdr = Buffer.alloc(13);
  ihdr.writeUInt32BE(size, 0);
  ihdr.writeUInt32BE(size, 4);
  ihdr[8] = 8;
  ihdr[9] = 6;
  const row = Buffer.alloc(1 + size * 4);
  const raw = Buffer.alloc((1 + size * 4) * size);
  for (let y = 0; y < size; y++) {
    row[0] = 0;
    for (let x = 0; x < size; x++) {
      const [r, g, b, a] = draw(x, y, size);
      const i = 1 + x * 4;
      row[i] = r;
      row[i + 1] = g;
      row[i + 2] = b;
      row[i + 3] = a;
    }
    row.copy(raw, y * row.length);
  }
  const compressed = deflateSync(raw);
  return Buffer.concat([
    signature,
    chunk('IHDR', ihdr),
    chunk('IDAT', compressed),
    chunk('IEND', Buffer.alloc(0)),
  ]);
}

function inRoundRect(x, y, size, rx, ry, rw, rh, rad) {
  const cx = Math.max(rx + rad, Math.min(x, rx + rw - rad - 1));
  const cy = Math.max(ry + rad, Math.min(y, ry + rh - rad - 1));
  const dx = x < rx + rad ? x - (rx + rad) : x >= rx + rw - rad ? x - (rx + rw - rad - 1) : 0;
  const dy = y < ry + rad ? y - (ry + rad) : y >= ry + rh - rad ? y - (ry + rh - rad - 1) : 0;
  if (Math.abs(x - cx) <= rad && Math.abs(y - cy) <= rad) return dx * dx + dy * dy <= rad * rad;
  return x >= rx && x < rx + rw && y >= ry && y < ry + rh;
}

function inHourglass(x, y, size) {
  const cx = size / 2;
  const cy = size * 0.46;
  const w = size * 0.22;
  const h = size * 0.34;
  const nx = (x - cx) / w;
  const ny = (y - cy) / h;
  if (Math.abs(nx) > 1) return false;
  const top = ny < 0 && Math.abs(ny) <= 1 - Math.abs(nx) * 0.15;
  const bot = ny > 0 && Math.abs(ny) <= 1 - Math.abs(nx) * 0.15;
  const neck = Math.abs(nx) < 0.12 && Math.abs(ny) < 0.08;
  return top || bot || neck;
}

function drawIcon(x, y, size) {
  const pad = size * 0.1;
  const rad = size * 0.18;
  if (!inRoundRect(x, y, size, pad, pad, size - pad * 2, size - pad * 2, rad)) {
    return [0, 0, 0, 0];
  }
  const stripeH = size * 0.1;
  const stripeY = size - pad - stripeH;
  if (y >= stripeY) {
    const third = (size - pad * 2) / 3;
    const lx = x - pad;
    if (lx < third) return [...VERDE, 255];
    if (lx < third * 2) return [...BIANCO, 255];
    return [...ROSSO, 255];
  }
  if (inHourglass(x, y, size)) return [...SAND, 255];
  return [...TERRA, 255];
}

function drawMaskable(x, y, size) {
  const cx = size / 2;
  const cy = size / 2;
  const r = size * 0.44;
  const dx = x - cx + 0.5;
  const dy = y - cy + 0.5;
  if (dx * dx + dy * dy > r * r) return [...BG, 255];
  return drawIcon(x, y, size);
}

for (const [name, size, fn] of [
  ['icon-192.png', 192, drawIcon],
  ['icon-512.png', 512, drawIcon],
  ['icon-512-maskable.png', 512, drawMaskable],
]) {
  writeFileSync(join(root, name), pngSolid(size, fn));
  console.log('OK', name);
}
